fish-rust/src/output.rs

```diff
@@ -8,12 +24,718 @@ bitflags! {
+            writembs!(self, &term.exit_attribute_mode);
+            self.last_color = normal;
+            self.last_color2 = normal;
+            self.reset_modes();
+        }
+        if !self.last_color2.is_special() {
+            // Background was set.
+            // "Special" here refers to the special "normal", "reset" and "none" colors,
+            // that really just disable the background.
+            last_bg_set = true;
+        }
+        if !bg.is_special() {
+            // Background is set.
+            bg_set = true;
+            if fg == bg {
+                fg = if bg == RgbColor::WHITE {
+                    RgbColor::BLACK
+                } else {
+                    RgbColor::WHITE
+                };
+            }
+        }
+
+        if term.enter_bold_mode.if_nonempty().is_some() {
+            if bg_set && !last_bg_set {
+                // Background color changed and is set, so we enter bold mode to make reading easier.
+                // This means bold mode is _always_ on when the background color is set.
+                writembs_nofail!(self, &term.enter_bold_mode);
+            }
+            if !bg_set && last_bg_set {
+                // Background color changed and is no longer set, so we exit bold mode.
+                writembs_nofail!(self, &term.exit_attribute_mode);
+                self.reset_modes();
+                // We don't know if exit_attribute_mode resets colors, so we set it to something known.
+                if write_foreground_color(self, 0, term) {
+                    self.last_color = RgbColor::BLACK;
+                }
+            }
+        }
+
+        if self.last_color != fg {
+            if fg.is_normal() {
+                write_foreground_color(self, 0, term);
+                writembs!(self, &term.exit_attribute_mode);
+
+                self.last_color2 = RgbColor::NORMAL;
+                self.reset_modes();
+            } else if !fg.is_special() {
+                self.write_color(fg, true /* foreground */);
+            }
+        }
+        self.last_color = fg;
+
+        if self.last_color2 != bg {
+            if bg.is_normal() {
+                write_background_color(self, 0, term);
+
+                writembs!(self, &term.exit_attribute_mode);
+                if !self.last_color.is_normal() {
+                    self.write_color(self.last_color, true /* foreground */);
+                }
+                self.reset_modes();
+                self.last_color2 = bg;
+            } else if !bg.is_special() {
+                self.write_color(bg, false /* not foreground */);
+                self.last_color2 = bg;
+            }
+        }
+
+        // Lastly, we set bold, underline, italics, dim, and reverse modes correctly.
+        let enter_bold_mode = term.enter_bold_mode.if_nonempty();
+        if is_bold && !self.was_bold && enter_bold_mode.is_some() && !bg_set {
+            // TODO: rationalize why only this one has the tparm0 call.
```

Comment:
Been that way since it was introduced in d66700a0e4d20a9193ea83490781c0858d98c3f6, no reason that i can discern.

I don't believe that tparm without any parameters is actually ever useful? As far as I can tell, the only thing that tparm does is to apply the given parameters, so if there's no parameter to apply it won't do anything.

Reply:
That is correct.

Reply:
I've removed tparm0.